drivers/accel/adxl355/adxl355.c

```diff
@@ -0,0 +1,934 @@
+		spi_remove(dev->spi_desc);
+	else
+		i2c_remove(dev->i2c_desc);
+	free(dev);
+	return ret;
+error_dev:
+	free(dev);
+	return ret;
+}
+
+/***************************************************************************//**
+ * @brief Free the resources allocated by adxl355_init().
+ *
+ * @param dev - The device structure.
+ *
+ * @return ret - Result of the remove procedure.
+*******************************************************************************/
+int adxl355_remove(struct adxl355_dev *dev)
+{
+	int ret;
+
+	if (dev->communication_type == ADXL355_SPI_COMM)
+		ret = spi_remove(dev->spi_desc);
+	else
+		ret = i2c_remove(dev->i2c_desc);
+
+	free(dev);
+
+	return ret;
+}
+
+/***************************************************************************//**
+ * @brief Reads from the device.
+ *
+ * @param dev          - The device structure.
+ * @param base_address - Address of the base register.
+ * @param size         - The number of bytes to be read and returned in read_data.
+ * @param read_data    - The read data buffer
+ *
+ * @return ret         - Result of the reading procedure.
+*******************************************************************************/
+int adxl355_read_device_data(struct adxl355_dev *dev, uint8_t base_address,
+			     uint16_t size, uint8_t *read_data)
+{
+	int ret;
+
+	if (dev->communication_type == ADXL355_SPI_COMM) {
+		uint8_t *comm_buffer;
+		comm_buffer = (uint8_t *)calloc(1,1 + size);
+		if (!comm_buffer)
+			return -ENOMEM;
+
+		comm_buffer[0] = ADXL355_SPI_READ | (base_address << 1);
+		ret = spi_write_and_read(dev->spi_desc, comm_buffer, 1 + size);
+		for (uint16_t idx = 0; idx < size; idx++)
+			read_data[idx] = comm_buffer[idx+1];
+
+		free(comm_buffer);
+	} else {
+		ret = i2c_write(dev->i2c_desc, &base_address, 1, 0);
+		if (ret)
+			return ret;
+		ret = i2c_read(dev->i2c_desc, read_data, size, 1);
+	}
+
+	return ret;
+}
+
+/***************************************************************************//**
+ * @brief Writes to the device
+ *
+ * @param dev          - The device structure.
+ * @param base_address - Address of the base register.
+ * @param size         - The number of bytes to be written. It is the size
+ *                       of the write_data buffer.
+ * @param write_data   - The data which is going to be written.
+ *
+ * @return ret         - Result of the writing procedure.
+*******************************************************************************/
+int adxl355_write_device_data(struct adxl355_dev *dev, uint8_t base_address,
+			      uint16_t size, uint8_t *write_data)
+{
+	int ret;
+	uint8_t *comm_buffer;
+	comm_buffer = (uint8_t *)calloc(1,1 + size);
+	if (!comm_buffer)
+		return -ENOMEM;
+	for (uint16_t idx = 0; idx < size; idx++)
+		comm_buffer[1+idx] = write_data[idx];
+
+	if (dev->communication_type == ADXL355_SPI_COMM) {
+		comm_buffer[0] = ADXL355_SPI_WRITE | (base_address << 1);
+		ret = spi_write_and_read(dev->spi_desc, comm_buffer, size + 1);
```

Comment:
also pay attention to endianess, multi-byte reads/writes may be influenced by endianess match/mismatch between the adxl355 and the MCU

Reply:
For multi-byte write, which is the case for the following APIs:
- adxl355_set_offset
- adxl355_configure_activity_threshold

The written data is split in bytes using shift operations so the functionality is not affected by endianess. 

For multi-byte read, which is the case for the following data:
- shadow registers value
- measured acceleration data
- temperature data
- fifo data

The read data is received in bytes, which are then assembled using shift operations, so the functionality is not affected by endianess.